### PR TITLE
[BREAKING] chore!: update redeem benchmarks

### PR DESCRIPTION
**Pull request by gregdhill**
Replaces `Formattable` and `TryFormattable` with `TryFormat` which now takes a `Writer` which can be bounded to stop formatting when the limit has been exceeded. Adds a `length_bound` to `execute_redeem` for the maximum encoded size of the Bitcoin transaction. This is seeded in the benchmark by extending the size of the input scripts. 

I have parameterized the following:
- number of hashes in the `MerkleProof` since computation is divisible by the number of branches in the tree
- number of transaction inputs since this affects transaction encoding
- number of transaction outputs (**currently ignored**)
- total size of the encoded transaction as described above

### CODE REVIEW

File: crates/bitcoin/src/types.rs
Comment (sander2) [inline crates/bitcoin/src/types.rs:325]: You can set `length_bound` and still execute all of `format_with` meaning the weight could get underestimated

File: crates/bitcoin/src/formatter.rs
Comment (sander2) [inline crates/bitcoin/src/formatter.rs:25]: I'm not sure I like having default implementation, it makes it easy to forget to properly implement for the various types

File: crates/bitcoin/src/types.rs
Comment (sander2) [inline crates/bitcoin/src/types.rs:319]: You probably won't want to do this but I think it'd be good to feature gate this(and `hash` below) so we can be sure we don't use it in parachain production

**Review by sander2 — COMMENTED** [review]
Could you add some tests?

File: crates/bitcoin/src/formatter.rs
Comment (gregdhill) [reply crates/bitcoin/src/formatter.rs:25]: Done, but it's not the most elegant solution. Let me know what you think

File: crates/bitcoin/src/types.rs
Comment (gregdhill) [reply crates/bitcoin/src/types.rs:319]: Agree but I did start to do this and it incurs a lot of changes, plus I removed some extrinsics in #1022 which use these methods so I would be inclined to wait until both PRs are merged then I can make the necessary changes to feature-gate those and the parser like you requested before.

File: crates/bitcoin/src/formatter.rs
Comment (sander2) [inline crates/bitcoin/src/formatter.rs:139]: To be pedantic, this technically is unbounded execution - you iterate over all elements of the vector without checking that the user paid weight for it. Unfortunately, fixing this requires another significant rewrite.. I can think of 2 ways to fix it:

(1) pass the length bound to `format_size` as well, or
(2) remove `format_size` as a trait method. I don't think this method it really necessary anyway - you can just call format_size, and have the call fail when it uses up its budget.

My preference would be 2

Comment (gregdhill) [reply crates/bitcoin/src/formatter.rs]: This is technically unused by the runtime, but I have refactored the formatting code to closely resemble that of `rust-bitcoin`. Using this approach we can pass a `BoundedWriter` to fail as soon as the budget has been consumed.

File: crates/bitcoin/src/formatter.rs
Comment (gregdhill) [reply crates/bitcoin/src/formatter.rs:13]: Indeed, see also my comment here: https://github.com/interlay/interbtc/pull/1020#discussion_r1177860775

File: crates/redeem/src/benchmarking.rs
Comment (sander2) [inline crates/redeem/src/benchmarking.rs:290]: I think this add `b * i` bytes, whereas we want to add only `b` bytes, right?

File: crates/redeem/src/benchmarking.rs
Comment (sander2) [inline crates/redeem/src/benchmarking.rs:230]: Just a note: I _think_ that we're counting some of this factors twice. For example, adding an input will increase `i`, but it will also increase `b`. But at worst we'll overestimate the weight, which I guess is acceptable

Comment (gregdhill) [reply crates/redeem/src/benchmarking.rs:230]: I had assumed the benchmarking would measure weights with respect to the relative change in each parameter. What I mean to say is for example I assume it would trial various settings of `b` when `i = x` to see how that affects computational and storage costs so the fact `i` may change the size of the Bitcoin transaction would not matter.

File: crates/redeem/src/benchmarking.rs
Comment (sander2) [reply crates/redeem/src/benchmarking.rs:290]: It's the variable `i` in the code, i.e. the number of inputs

File: crates/redeem/src/benchmarking.rs
Comment (sander2) [reply crates/redeem/src/benchmarking.rs:230]: It's not quite clear to me how it works: https://github.com/paritytech/substrate/blob/master/frame/benchmarking/src/analysis.rs

But worst case scneario it's just overkill, which is fine

**Review by sander2 — APPROVED** [review]

**Pull request merged**
